--- Portage.pm
+++ Portage.pm
@@ -205,7 +205,8 @@
 
 sub read_profiles() {
 	$_ = readlink '/etc/make.profile';
-	die "/etc/make.profile is not a symlink\n" if not defined $_;
+	$_ = readlink '/etc/portage/make.profile' if not defined $_;
+	die "/etc\{,/portage\}/make.profile is not a symlink\n" if not defined $_;
 	@profiles = norm_path '/etc', $_;
 	for (my $i = -1; $i >= -@profiles; $i--) {
 		for(noncomments "$profiles[$i]/parent") {
