--- 1/ccache-config
+++ 1/ccache-config
@@ -66,15 +66,17 @@
 }
 cc_links() {
 	local a
-	for a in gcc cc c++ g++ ; do
-		if [ -n "${2}" ] ; then
-			# gcc-config doesnt install ${CHOST}-cc, so until
-			# it does, don't install a ccache symlink for it
-			[ "${a}" = "cc" ] && continue
-			a="${2}-${a}"
-		fi
-		eval "cc_${1}_link" "${a}"
-	done
+	if [ -z "${2}" ] ; then
+		# gcc-config doesn't install ${CHOST}-cc, so until
+		# it does, don't install a ccache symlink for it
+		for a in cc clang clang++
+		do	eval "cc_${1}_link" "${a}"
+		done
+	else
+		for a in gcc c++ g++
+		do	eval "cc_${1}_link" "${2}-${a}"
+		done
+	fi
 }
 
 ###
